src/data/CCRAL.ml

```diff
@@ -116,11 +116,22 @@ let front_exn l = match l with
 let rec _remove prefix l i =
   let x, l' = front_exn l in
   if i=0
-  then List.fold_left (fun l x -> cons x l) l prefix
+  then List.fold_left (fun l x -> cons x l) l' prefix
   else _remove (x::prefix) l' (i-1)
 
 let remove l i = _remove [] l i
 
+let rec _get_and_remove_exn found prefix l i =
+  let x, l' = front_exn l in
+  if i=0
+  then (found := Some x; List.fold_left (fun l x -> cons x l) l' prefix)
+  else _get_and_remove_exn found (x::prefix) l' (i-1)
+
+let get_and_remove_exn l i =
+  let found = ref None in
+  let l' = _get_and_remove_exn found [] l i in
+  (Option.get !found, l')
```

Comment:
```suggestion
let rec _get_and_remove_exn prefix l i =
  match front l with
  | None -> None, l
  | Some (x,l') ->
    if i=0
    then (Some x, List.fold_left (fun l x -> cons x l) l' prefix)
    else _get_and_remove_exn found (x::prefix) l' (i-1)

let get_and_remove_exn l i =
  _get_and_remove_exn found [] l i
```

this would be more direct and would avoid the ref (just return a tuple directly), I believe.

Reply:
Yes, I'm guilty of premature optimization by default here.